CodeEdit/Features/DependencyInjection/ServiceContainer.swift

```diff
@@ -0,0 +1,44 @@
+//
+//  ServiceContainer.swift
+//  CodeEdit
+//
+//  Created by Abe Malla on 4/3/24.
+//
+
+enum ServiceContainer {
+    private static var factories: [String: () -> Any] = [:]
+    private static var cache: [String: Any] = [:]
+
+    static func register<Service>(type: Service.Type, _ factory: @autoclosure @escaping () -> Service) {
+        factories[String(describing: type.self)] = factory
+    }
+
+    static func resolve<Service>(_ resolveType: ServiceType = .singleton, _ type: Service.Type) -> Service? {
+        let serviceName = String(describing: type.self)
+
+        switch resolveType {
+        case .singleton:
+            if let service = cache[serviceName] as? Service {
+                return service
+            } else {
+                let service = factories[serviceName]?() as? Service
+
+                if let service = service {
+                    cache[serviceName] = service
+                }
+
+                return service
+            }
+        case .newSingleton:
+            let service = factories[serviceName]?() as? Service
+
+            if let service = service {
+                cache[serviceName] = service
+            }
+
+            return service
+        case .new:
+            return factories[serviceName]?() as? Service
```

Comment:
Should we return an error like 
```swift
throw ServiceError.serviceNotRegistered
```

Reply:
An error is thrown inside the Service property wrapper, so this isn't needed